### ISSUE TITLE
Sequence is copied to closure, which is slow

### ISSUE THREAD

**Issue by petermora** [nim-lang/Nim#3377]
Hi,

 In the following example the sequence ``x`` is copied for the closure. Is this intended? Copying is very slow.

```nim
import sequtils

proc toIter[T](s: seq[T]): (iterator(): T) {.inline.} =
  result = iterator(): T {.closure.}=
    for elem in s:
      yield elem

var x = toSeq(1..1000)
var iter = toIter(x)  # In this line there is a copy of x.
echo iter()
```

This is the C code:
```c
N_NIMCALL(void, toiter_95093)(TY95022* s, TY95099* Result) {
        TY95218* HEX3Aenv_95227;
        TY95099 LOC1;
        HEX3Aenv_95227 = 0;
        HEX3Aenv_95227 = (TY95218*) newObj((&NTI95225), sizeof(TY95218));
// The next line copies s variable to the closure.
        genericSeqAssign((&(*HEX3Aenv_95227).s95098), s, (&NTI95022));
        memset((void*)(&LOC1), 0, sizeof(LOC1));
        LOC1.ClPrc = HEX3Aanonymous_95101; LOC1.ClEnv = HEX3Aenv_95227;
        unsureAsgnRef((void**) (&(*Result).ClEnv), LOC1.ClEnv);
        (*Result).ClPrc = LOC1.ClPrc;
}
```
The ``genericSeqAssign(...)`` proc calls ``genericAssign(...)`` in *assign.nim:132*.

I can speed up my bechmarks (and have the right result) if I don't care about the consequences and modify the last parameter from ``false`` to ``true`` in *assign.nim:95*, therefore creating only a shallow copy:
```nim
 proc genericAssign(dest, src: pointer, mt: PNimType) {.compilerProc.} =
    genericAssignAux(dest, src, mt, false)
```

Could you please advise?

Thank you,
  Peter

**Comment by petermora** [comment]
Let me provide an additional test:
```nim
proc giveMeFunction(s: seq[int]): proc() =
  result = proc() =
    echo s

let mySeq = newSeq[int](10_000_000)
#var myFun = giveMeFunction(mySeq)
echo GC_getStatistics()
```
This code prints: [GC] total memory: **84**201472, which feels right, because I have a 64 bit linux, and 10MM * 8 bytes, that's roughly 80MM bytes.

If I enable the ``myFun`` line, then the memory usage jumps to: [GC] total memory: **164***204544.

I might be wrong about this, I thought that if I'm not doing explicitly ``var x = mySeq``, then there should be no memory copy.

Thanks,
 Peter

**Comment by petermora** [comment]
It looks like that the ``newAsgnStmt(...)`` function copies the variable in *lambdalifting.nim*. If I change the ``nkAsgn`` to ``nkFastAsng`` (and do some tricks to compile the compiler), then my sequence is not copied as I wanted. For the string literals and the sequence of string literals the ``nkFastAsng`` is broken as it is stated by a comment in ``newAsgnStmt(...)``, I don't have a solution.

**Comment by Varriount** [comment]
Um, isn't this expected behavior? Sequences have deep copying semantics, so that they only ever have a refcount of one

**Comment by petermora** [comment]
The motivation behind this issue is that I would like to have a fast iterator library. The first function to have in this library is the ``toIter(...)`` proc above. And it is amazingly slow currently.

My intuition was that if I don't write assignments like ``var x = mySeq``, only call functions like ``f(mySeq)``, then there is no copy. Therefore I would assume that there is no copy in the example above. I understand your argument about sequences should have refcount of one. However, we have ``let`` and ``shallowCopy``, which avoids copying.

Currently the best (and only) solution I can see: we could test whether the current variable has a GC header (I don't know how to do it, but I guess that it is not that hard to test, string literals have a very different memory address). And in ``newAsgnStmt(...)`` function we could generate a code which will test this in runtime and run either ``nkAsgn`` or ``nkFastAsgn`` depending on the result. (I can't see a solution how to do this in compile time.)

Jiust a side note, I don't want to hijack the thread: I haven't mentioned that if one of the argument is ``var`` like: ``proc toIter[T](s: var seq[T]):...``, then I get *illegal capture of 's'* compile time exception. We could start a long discussion whether this would be healthy to do.

**Comment by petermora** [comment]
Should I continue to work on this (to avoid copy when it is possible), or just leave it as it is? Is there any reason for the current behavior?

**Comment by Varriount** [comment]
If there's a way to fix this, then do so by all means. If there's one big improvement Nim's built-in sequence and string types need, it's optimization with regards to copying behavior.

That being said, using both `let` and `shallowCopy` preclude modifying and extending the target sequence, respectively.  
To use non-copying semantics with sequences when creating a closure iterator, the compiler will need to ensure that the sequence is neither modified nor garbage collected during the lifetime of the closure.

Perhaps a long term goal would be a string/sequence analysis pass which analyses the role each possible string and sequence type plays in a program.

**Comment by Araq** [comment]
> Should I continue to work on this (to avoid copy when it is possible), or just leave it as it is? Is there any reason for the current behavior?

Yes. Memory safety. You need to **ensure** incRef is not performed on static data that's not on the heap.

**Comment by petermora** [comment]
Let me ask it different way. If I can (with some magic) ensure that refcount is not increased if the data is not on heap (but copied instead), then can I do that? The following behavior will change:
```nim
proc giveMeFun(s: seq[int]): proc() =
  result = proc() =
    echo s[0]
var seqWillChange = newSeq[int](3)
var f = giveMeFun(seqWillChange)
seqWillChange[0] = 99
f() # prints 0 now, and will print 99 after the fix
```
I'm completely ok with this. I don't want to ensure that the seq is not modified the lifetime of the closure, but maybe that's only me. That's why I'm asking. (And I couldn't ensure that the seq is not modified even if I wanted.)

Cheers, Peter

**Comment by Araq** [comment]
> Let me ask it different way.

You can do that and I'm in general happy with everything that improves performance. However, it might be a better idea to join me on IRC and discuss a general solution to the problem. I wrote a **very** powerful escape analysis that can make these transformations safe and general. I'm very busy with fixing bugs though, not with optimizations, so if you're willing to help, this would be great. The code is in compiler/writetracking.nim and disabled for now.

**Comment by petermora** [comment]
Same topic, so let me paste it here instead of starting a new issue:
```nim
proc f(x: seq[int]) =
  var y: seq[int]
  shallowCopy(y, x)
  y[0] = 99    # throws SIGSEGV: Illegal storage access.

f(@[1,2,3])
```

**Comment by petermora** [comment]
The following example shows that it is not enough to decide whether the seq is in static or on the heap (because the output is different). Some additional smart compiler logic is needed.
```nim
proc smartCopyOrShallowCopy(dest: var seq[int], src: seq[int]) =
  # This proc works in this example on my 64-bit linux, but not portable!!!
  if cast[int](cast[pointer](src)) < 9999999:
    dest = src     # src is in static
  else:
    shallowCopy(dest, src)       # src is on the heap

proc f(x: seq[int]) =
  var y: seq[int]
  smartCopyOrShallowCopy(y, x)
  y[0] = 99
  echo x[0] # prints either 99 or 1

var q = @[1, 2, 3]
f(q)  # prints 99

f(@[1,2,3]) # prints 1
```
 It looks like ``shallowCopy`` can be a dangerous proc.

**Comment by Araq** [comment]
@petermora We know this. shallowCopy is declared as unsafe for these reasons.

**Comment by petermora** [comment]
@Araq I assumed that ``shallowCopy`` is safe to use, I didn't see any unsafe declaration. It is not pointer magic and it is in the core lib: system.nim. If we consider it unsafe, then could we call it ``unsafeShallowCopy``?

Also do you think that ``shallowCopy(x: var T, y: T)`` should be modified to ``shallowCopy(x: var T, y: var T)``? The reason is simple: if you have a read-only ``y``, then it feels cheating to create a writable copy of it. 

We could have both (``shallowCopy(var T, var T)`` which is now safe and can't have static variable as input, and ``unsafeShallowCopy(var T, T)`` which warns the user by its name). What do you think?

I still think about the lambda lifting, playing with shallowCopy. I keep in mind your escape detection mechanism (although didn't have time to check the code).

**Comment by Araq** [comment]
> We could have both (shallowCopy(var T, var T) which is now safe and can't have static variable as input, and unsafeShallowCopy(var T, T) which warns the user by its name). What do you think?

Yes, that's an excellent solution that was also proposed by jehan.

**Comment by petermora** [comment]
@Araq Could you please comment on these ideas? I'm asking this because 10 minutes of your time can save hours or days from mine.

 The problem (again):
```nim
proc giveMeProc(s: seq[int]): proc() =
  result = proc()
    echo s[0]
var x = someLargeSeq()
var f = giveMeProc(x)
f()
```
This code copies ``x``, which is slow and should be only a reference copy. The motivation is to build a proper iterator library.

In *lambdalifting.nim* the ``newAsgnStmt(...)`` uses ``nkAsgn`` and not ``nkFastAsgn``. If we change it to ``nkFastAsgn`` then the following code tries to increase a GC counter on static data, which is segfault:
```nim
var f = giveMeProc(@[1,2,3])
```
I'm writing about sequences all the time, but everything here applies to strings as well.

Let me list 3 different solutions which is safe (no write on static data) and fast (no copy of large list). The order is based on how hard to implement them. I'm focusing here on C, and have a memory layout in my head described here: https://en.wikipedia.org/wiki/Data_segment


**Solution 1.** (easiest to implement)
Let's keep a pointer to the lowest and highest heap memory address where GC is storing data. Then in ``newAsgnStmt(...)`` we could implement a runtime check which copies if the memory is not on the heap, otherwise it's just a reference copy. We need to understand what happens with multithreads: if the sequence is managed by a different thread's garbage collector. The object's reference counter is not thread-safe, so I guess that we need to copy in these situations (just imagine that multiple threads are increasing/decreasing the reference counter). I can see two possible workarounds:
- We do a deep copy (``nkAsgn``) if it is a multithreading environment. This is a half solution, but better than nothing.
- We do a proper check of the current thread's heap: if ``--threads:on``, then we don't only compare the seq's memory address to the largest and lowest heap memory addresses, but also iterate over the memory regions to ensure that this thread has this variable (for each memory region we could store the largest and lowest memory address).

**Solution 2.**
In C's memory layout, the Uninitialized Data (BSS) and the Initialized Data are stored at fix memory location and they are shared among the threads. You can list their size with a linux command "size your_binary". I don't know how we could get their memory address in either compile time or runtime. However, with the addresses of the boundaries of these memory segments it would be an easy runtime check whether the given pointer is on static.

**Solution 3.**
In the example above we could mark ``s`` as lifted variable. This would change the signature of ``giveMeProc`` from ``giveMeProc(s: seq[int]): proc()`` to ``giveMeProcs: mustBeOnTheHeap seq[int]): proc()``. We also need to keep track of the variables: the ``giveMeProc(@[1,2,3])`` is a call with parameter which is in static, therefore we need to copy it before the function call.
This solution is not that easy to implement and also has some bad implications. In the following case:
```nim
proc wrapper(s: seq[int]): proc() =
  result = giveMeProc(s)
```
we need to realize that the parameter of ``wrapper`` needs to be on the heap. Moreover, if at the implementation at ``wrapper`` we might not now the actual implementation of ``giveMeProc``, therefore we actually need to introduce an additional keyword to the language (to indicate that the parameter ``s`` needs to be on the heap at forward declarations).

Any comments are welcome.

Peter

**Comment by Araq** [comment]
Solution 3 is essentially:  

proc wrapper(s: seq[int]): proc () {.escapes: [s].} 

That's why I told you to look at my writetracking. ;-)
Escapes: [s]  means you cannot pass a static seq to it.

**Comment by Varriount** [comment]
Is `escapes` an actual pragma?

**Comment by petermora** [comment]
@Araq I see. If you don't write ``escapes`` pragma for a proc, then will lambda lifting throw compile error?

The ``escapes`` pragma will make the code easier to understand, it will help the reader. However, we require information from the coder which could be calculated by the compiler. I think this as an optimal solution, but I can live with that. I really should start to read your escape analysis code.

**Comment by Araq** [comment]
@petermora Well 'escapes' is an effect and as such is inferred for you. The usual effect tracking rules apply though (indirect calls and forward decls require explicit effect annotations).
